### PR TITLE
HIR-19. Add endpoints to manage contract rates.

### PR DESCRIPTION
**Pull request by oleg1std**

### CODE REVIEW

File: HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        }
+
+
+        public Task<Result<List<Models.Responses.Rate>>> Add(int contractId, List<Models.Requests.Rate> rates)
+        {
+            return ValidationHelper.Validate(rates, new RateValidator())
+                .Bind(() => _contractManagerContext.GetContractManager())
+                .Ensure(contractManager => DoesContractBelongToContractManager(contractId, contractManager.Id),
+                    $"Failed to get the contract by {nameof(contractId)} '{contractId}'")
+                .Tap(async contractManager => Result.Combine(await CheckIfSeasonsBelongToContract(contractId, rates.Select(rate => rate.SeasonId).ToList()),
```
Comment (kirillta) [inline HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs:42]: This `Tap` looks over-complicated

File: HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+            var ratesAndRoomsAndSeasons = _dbContext.RoomRates
+                .Join(_dbContext.Rooms, roomRate => roomRate.RoomId, room => room.Id, (roomRate, room) => new
+                    {roomRate, room})
+                .Join(_dbContext.Seasons, rateAndRoom => rateAndRoom.roomRate.SeasonId, season => season.Id, (roomAndRate, season) => new
+                    {roomAndRate.roomRate, roomAndRate.room, season})
+                .Where(rateAndRoomAndSeason => contractedAccommodationIds.Contains(rateAndRoomAndSeason.room.AccommodationId))
+                .Where(rateAndRoomAndSeason => rateAndRoomAndSeason.season.ContractId == contractId);
+            
+            if (roomIds != null && roomIds.Any())
```
Comment (mingazhev) [inline HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs:77]: Please add the curly brackets there, this is hard to read for me

File: HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                .Ensure(contractManager => DoesContractBelongToContractManager(contractId, contractManager.Id),
+                    $"Failed to get the contract by {nameof(contractId)} '{contractId}'")
+                .Bind(contractManager => CheckAndGetRatesToRemove(contractId, contractManager.Id, rateIds))
+                .Bind(RemoveRates);
+        }
+
+
+        private async Task<List<RoomRate>> GetRates(int contractId, int contractManagerId, List<int> roomIds = null, List<int> seasonIds = null)
+        {
+            var contractedAccommodationIds = _dbContext.Accommodations
```
Comment (mingazhev) [inline HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs:60]: Wouldn't this be more readable if was written using LINQ query syntax?

File: HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                ratesAndRoomsAndSeasons = ratesAndRoomsAndSeasons.Where(rateAndRoomAndSeason => roomIds.Contains(rateAndRoomAndSeason.room.Id));
+
+            if (seasonIds != null && seasonIds.Any())
+                ratesAndRoomsAndSeasons = ratesAndRoomsAndSeasons.Where(rateAndRoomAndSeason => seasonIds.Contains(rateAndRoomAndSeason.season.Id));
+
+            return await ratesAndRoomsAndSeasons.Select(i => i.roomRate).Distinct().ToListAsync();
+        }
+
+
+        private async Task<Result<List<RoomRate>>> CheckAndGetRatesToRemove(int contractId, int contractManagerId, List<int> rateIds)
```
Comment (mingazhev) [inline HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs:87]: Do we need a result there? Why not returning an empty collection?

File: HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+            var checkingResult = Result.Combine(await CheckIfSeasonsBelongToContract(contractId, roomRates.Select(roomRate => roomRate.SeasonId).ToList()),
+                await CheckIfRoomIdsBelongToContractedAccommodation(contractId, contractManagerId, roomRates.Select(roomRate => roomRate.RoomId).ToList()));
+
+            return checkingResult.IsFailure
+                ? Result.Failure<List<RoomRate>>(checkingResult.Error)
+                : Result.Success(roomRates);
+        }
+
+
+        private async Task<Result> RemoveRates(List<RoomRate> rates)
```
Comment (mingazhev) [inline HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs:102]: Is Result is needed there? There is no unsuccessful paths

File: HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs (hunk 4 above)
Comment (oleg1std) [reply HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs:87]: There are additional checks for rates to delete

File: HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs (hunk 3 above)
Comment (oleg1std) [reply HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs:60]: It's more readable in my opinion. 
Another one topic to discuss

File: HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+
+        public Task<Result<List<Models.Responses.Rate>>> Add(int contractId, List<Models.Requests.Rate> rates)
+        {
+            return ValidationHelper.Validate(rates, new RateValidator())
+                .Bind(() => _contractManagerContext.GetContractManager())
+                .Ensure(contractManager => DoesContractBelongToContractManager(contractId, contractManager.Id),
+                    $"Failed to get the contract by {nameof(contractId)} '{contractId}'")
+                .Tap(contractManager => CheckIfSeasonsBelongToContract(contractId, rates.Select(rate => rate.SeasonId).ToList()))
+                .Tap(contractManager => CheckIfRoomIdsBelongToContractedAccommodation(contractId, contractManager.Id, rates.Select(rate => rate.RoomId).ToList()) )
```
Comment (kirillta) [inline HappyTravel.Hiroshima.DirectManager/Services/RateManagementService.cs:43]: Those `Tap`s look like you not interesting in check results. You add rates despite check results

**Pull request merged**